content/post/2020-01-02-cran-error/index.Rmd

```diff
@@ -0,0 +1,101 @@
+
+- [rhub](https://r-hub.github.io/rhub/), which is an R package for using the r-hub,
+- and the [Rocker project](https://www.rocker-project.org/), which distributes various docker images useful for R and RStudio testing
+
+If you are a maintainer of an R package and don't want to install docker locally, you will find the `rhub` very useful, as you can push your package to the server for testing with one command! Check out e.g., [this blog post](https://blog.r-hub.io/2019/04/25/r-devel-linux-x86-64-debian-clang/)
+
+If you want to test more locally, and importantly, if you want to test visualizations, GUI, or specific behavior of the code in RStudio, you should take a look on the `Rocker project`. They provide easy-to-run docker images with pre-installed R and RStudio, which you can play with through a browser window.
+
+## R-hub locally
+
+In my case, I am not registered as the main maintainer of Haplin, and I wanted to test locally, so I decided to use one of the `rhub` functions, [`local_check_linux()`](https://r-hub.github.io/rhub/reference/local_check_linux.html). This seemed like a very easy way to run locally tests on a platform that is exactly the same as CRAN uses! And it is... only not in my case ;)
+
+After getting the same error for the nth time, I decided to get help [at the source](https://github.com/r-hub/rhub/issues/322). I tried several things, including digging into the `rhub` code locally, but nothing worked. Long story short - my docker installation by default gave no access to internet to the containers it launched.
+
+:sparkles: Tip \#3 :sparkles: *ask for help other people!* (they don't bite, especially not via e-mails)
+
+## Script for semi-manual checks
+
+Another couple of days and I talked with my husband, who showed me a script he used in one of his projects. [The script](https://bitbucket.org/Grantlab/bio3d/src/master/ver_devel/util/run_dockercheck.sh) uses a local docker installation, fetches one of the images hosted by `rhub`, and runs tests locally. *Bingo!* As this was the second time I tried using docker, I needed some attempts to adjust the script to my purposes, but it finally worked!
+
+## Conclusion
+
+Thanks to `rhub` and other tools available online, it's relatively easy to test a CRAN package.
+
+
+### _Post scriptum_
+
+While writing this post, I've come upon another R package [dockr](https://cran.r-project.org/package=dockr), which helps creating a docker container with the package one wants to test already available within the container. I haven't tried it yet, but [it seems very promising](http://smaakage85.netlify.com/2019/12/21/dockr-easy-containerization-for-r/)!
+
+
+> Find Julia on [github](https://github.com/jromanowska), [bitbucket](https://bitbucket.org/jrom/profile/repositories), or [facebook](https://www.facebook.com/julia.romanowska.733)
+
+
+[^1]: I repeatedly refuse to use google - check out [here](https://spreadprivacy.com/how-to-remove-google/)
```

Comment:
```suggestion
[^1]: I repeatedly refuse to use Google [to protect my privacy](https://spreadprivacy.com/how-to-remove-google/)
```

Reply:
I've been told that one shouldn't use "click here" links https://webaccess.berkeley.edu/ask-pecan/click-here

Reply:
In case you don't know this package, you might enjoy it: https://github.com/dirkschumacher/duckduckr

Reply:
Thanks for the tips!